src/main/java/jcifs/smb1/ntlmssp/Type3Message.java

```diff
@@ -0,0 +1,655 @@
+        return sessionKey;
+    }
+
+    /**
+     * Sets the session key.
+     *
+     * @param sessionKey The session key.
+     */
+    public void setSessionKey(byte[] sessionKey) {
+        this.sessionKey = sessionKey;
+    }
+
+    public byte[] toByteArray() {
+        try {
+            int flags = getFlags();
+            boolean unicode = (flags & NTLMSSP_NEGOTIATE_UNICODE) != 0;
+            String oem = unicode ? null : getOEMEncoding();
+            String domainName = getDomain();
+            byte[] domain = null;
+            if (domainName != null && domainName.length() != 0) {
+                domain = unicode ?
+                        domainName.getBytes(UNI_ENCODING) :
+                                domainName.getBytes(oem);
+            }
+            int domainLength = (domain != null) ? domain.length : 0;
+            String userName = getUser();
+            byte[] user = null;
+            if (userName != null && userName.length() != 0) {
+                user = unicode ? userName.getBytes(UNI_ENCODING) :
+                        userName.toUpperCase().getBytes(oem);
+            }
+            int userLength = (user != null) ? user.length : 0;
+            String workstationName = getWorkstation();
+            byte[] workstation = null;
+            if (workstationName != null && workstationName.length() != 0) {
+                workstation = unicode ?
+                        workstationName.getBytes(UNI_ENCODING) :
+                                workstationName.toUpperCase().getBytes(oem);
+            }
+            int workstationLength = (workstation != null) ?
+                    workstation.length : 0;
+            byte[] lmResponse = getLMResponse();
+            int lmLength = (lmResponse != null) ? lmResponse.length : 0;
+            byte[] ntResponse = getNTResponse();
+            int ntLength = (ntResponse != null) ? ntResponse.length : 0;
+            byte[] sessionKey = getSessionKey();
+            int keyLength = (sessionKey != null) ? sessionKey.length : 0;
+            byte[] type3 = new byte[64 + domainLength + userLength +
+                    workstationLength + lmLength + ntLength + keyLength];
+            System.arraycopy(NTLMSSP_SIGNATURE, 0, type3, 0, 8);
+            writeULong(type3, 8, 3);
+            int offset = 64;
+            writeSecurityBuffer(type3, 12, offset, lmResponse);
+            offset += lmLength;
+            writeSecurityBuffer(type3, 20, offset, ntResponse);
+            offset += ntLength;
+            writeSecurityBuffer(type3, 28, offset, domain);
+            offset += domainLength;
+            writeSecurityBuffer(type3, 36, offset, user);
+            offset += userLength;
+            writeSecurityBuffer(type3, 44, offset, workstation);
+            offset += workstationLength;
+            writeSecurityBuffer(type3, 52, offset, sessionKey);
+            writeULong(type3, 60, flags);
+            return type3;
+        } catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)